### ISSUE TITLE
Problem adding to google home using nginx.

### ISSUE THREAD

**Issue by landowark** [mikejac/node-red-contrib-google-smarthome#276]
I'm having a hard time with the initial addition of the configuration with google home. As far as I can tell I've gotten everything set up correctly until it comes to linking the app into google home on android. The home app gives me "unable to connect" and the debug log on node red reads:

Auth:isValidClient(): clientSecret does not match (expected "\<string I have set as client secret\>", got "\<string I have set as client ID\>")!

As far as I can tell I have the clientsecret and clientID values set in the correct places in both the node red config and the google actions console.

The thing is if I go to the check url on my LAN at http://\<my local ip\>:3001/check and put in the values I get the following:

![image](https://user-images.githubusercontent.com/14314236/160242453-9144585b-bca7-4116-8440-ea537eddd969.png)

If I do the same using the WAN connection (I have a reverse proxy set up with Nginx) I get:

![image](https://user-images.githubusercontent.com/14314236/160242802-ae70ed42-1059-4c02-a77e-ca0c1f24f26f.png)

So, I'm guessing there is a problem with my nginx configuration:

```
server {
    server_name <my domain>;
    
    ssl_certificate <path to ssl_certificate>;
    ssl_certificate_key <path to ssl_certificate_key>;

    ssl_dhparam /etc/nginx/ssl/dhparams.pem;

    listen 443 ssl;
    add_header Strict-Transport-Security "max-age=31536000; includeSubdomains";

    ssl_protocols TLSv1.2;
    ssl_ciphers "EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH:!aNULL:!eNULL:!EXPORT:!DES:!MD5:!PSK:!RC4";
    ssl_prefer_server_ciphers on;
    ssl_session_cache shared:SSL:10m;

    proxy_buffering off;

    location / {
        proxy_pass http://localhost:3001/;
        proxy_set_header Host $host;
        proxy_redirect http:// https://;
        proxy_http_version 1.1;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
    }
}
```

Has anyone else been able to solve this problem with using nginx as a reverse proxy?

**Comment by Caprico85** [comment]
I need to experiment a bit more. But it could be a problem that you are proxying from port 443 to port 3001. node-red-contrib-google-smarthome thinks it is running on port 3001. So URLs are validated against port 3001, but URLs coming from external have port 443.

<br>

I'm not sure if the problem with client ID and secret is related to nginx.

> Auth:isValidClient(): clientSecret does not match (expected "\<string I have set as client secret>", got "\<string I have set as client ID>")!

It seems something is mixing up client id and secret. I'm not using nginx for my smarthome (Caddy as reverse proxy here). But I'm using nginx at work to load balance a website. I never had seen nginx messing request parameters. So I have to ask this: Are you really sure client id and client secret are
set correctly in Node-RED and in Actions on Google Console?

**Comment by landowark** [comment]
I double checked the ID and secret and now it's working. I swear they were correct before, but I've screwed up more obvious things before. Doesn't explain why there was success on the local ip and failure through nginx, but whatever, it's working. Thanks for the help.

**Comment by Caprico85** [comment]
I was able to reproduce this. As expected, it was related to using different internal and external ports. During its simulated login, the check page sends a redirect URL `https://example.com` (port 443 omitted as it is the default port). On server side, the URL is compared to `https://example.com:3001`. These URLs are not identical, so the check fails. As using different ports is fine, the comparison is now done without the ports. The fix will be included in the next release.

This only affected the simulated login on the check page. During a real login Google uses a totally different redirect URL (https://oauth-redirect.googleusercontent.com). Therefore the problem didn't occur during a real login.

**Comment by Caprico85** [comment]
Fix released in v0.3.2.
